File: app/src/main/java/it/amonshore/comikkua/ui/comics/ComicsFragment.java
package it.amonshore.comikkua.ui.comics;

import android.content.Context;
import android.os.Bundle;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;

import com.bumptech.glide.Glide;
import com.google.android.material.snackbar.BaseTransientBottomBar;
import com.google.android.material.snackbar.Snackbar;

import java.util.Objects;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.view.ActionMode;
import androidx.appcompat.widget.SearchView;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;
import androidx.navigation.NavDirections;
import androidx.navigation.Navigation;
import androidx.recyclerview.selection.SelectionTracker;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import it.amonshore.comikkua.LogHelper;
import it.amonshore.comikkua.R;
import it.amonshore.comikkua.Utility;
import it.amonshore.comikkua.data.comics.ComicsViewModel;
import it.amonshore.comikkua.data.comics.ComicsWithReleases;
import it.amonshore.comikkua.ui.ActionModeController;
import it.amonshore.comikkua.ui.ImageHelper;
import it.amonshore.comikkua.ui.OnNavigationFragmentListener;

import static it.amonshore.comikkua.data.comics.Comics.NEW_COMICS_ID;


public class ComicsFragment extends Fragment {

    private final static String BUNDLE_COMICS_RECYCLER_LAYOUT = "bundle.comics.recycler.layout";
    private final static String BUNDLE_COMICS_LAST_QUERY = "bundle.comics.last.query";

    private OnNavigationFragmentListener mListener;
    private PagedListComicsAdapter mAdapter;
    private ComicsViewModel mComicsViewModel;
    private RecyclerView mRecyclerView;
    private Snackbar mUndoSnackBar;

    public ComicsFragment() { }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setHasOptionsMenu(true);
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        final View view = inflater.inflate(R.layout.fragment_comics, container, false);

        final String actionModeName = getClass().getSimpleName() + "_actionMode";
        final Context context = requireContext();
        mRecyclerView = view.findViewById(R.id.list);
        mRecyclerView.setLayoutManager(new LinearLayoutManager(context));

        final ActionModeController actionModeController = new ActionModeController(R.menu.menu_comics_selected) {
            @Override
            public boolean onActionItemClicked(ActionMode mode, MenuItem item) {
                if (item.getItemId() == R.id.deleteComics) {
                    final SelectionTracker<Long> tracker = mAdapter.getSelectionTracker();
                    if (tracker.hasSelection()) {
                        // prima elimino eventuali release ancora in fase di undo
                        mComicsViewModel.deleteRemoved();
                        mComicsViewModel.remove(tracker.getSelection(), (ids, count) -> showUndo(ids, count));
                    }
                    tracker.clearSelection();
                    return true;
                }
                return false;
            }

            @Override
            public void onDestroyActionMode(ActionMode mode) {
                // action mode distrutta (anche con BACK, che viene gestito internamente all'ActionMode e non può essere evitato)
                mAdapter.getSelectionTracker().clearSelection();
                super.onDestroyActionMode(mode);
            }
        };

        // PROBLEMI:
        // selection changed viene scatenato due volte all'inizio: questo perché il tracker permette la selezione di più item trascinando la selezione

        mAdapter = new PagedListComicsAdapter.Builder(mRecyclerView)
                .withOnItemSelectedListener((keys, size) -> {
                    if (mListener != null) {
                        if (size == 0) {
                            mListener.onFragmentRequestActionMode(null, actionModeName, null);
                        } else {
                            mListener.onFragmentRequestActionMode(actionModeController, actionModeName,
                                    getString(R.string.title_selected, size));
                        }
                    }
                })
                .withComcisCallback(new PagedListComicsAdapter.ComicsCallback() {
                    @Override
                    public void onComicsClick(@NonNull ComicsWithReleases comics) {
                        final NavDirections directions = ComicsFragmentDirections
                                .actionDestComicsToComicsDetailFragment()
                                .setComicsId(comics.comics.id);

                        Navigation.findNavController(requireView()).navigate(directions);
                    }

                    @Override
                    public void onNewRelease(@NonNull ComicsWithReleases comics) {
                        final NavDirections directions = ComicsFragmentDirections
                                .actionDestComicFragmentToReleaseEditFragment()
                                .setComicsId(comics.comics.id)
                                .setSubtitle(R.string.title_release_create);

                        Navigation.findNavController(requireView()).navigate(directions);
                    }
                })
                .withGlide(Glide.with(this))
                .build();

        // recupero il ViewModel per l'accesso ai dati
        // lo lego all'activity perché il fragment viene ricrecato ogni volta (!)
        mComicsViewModel = new ViewModelProvider(requireActivity())
                .get(ComicsViewModel.class);
        // mi metto in ascolto del cambiamto dei dati (via LiveData) e aggiorno l'adapter di conseguenza
        mComicsViewModel.comicsWithReleasesList.observe(getViewLifecycleOwner(), data -> {
            LogHelper.d("comics viewmodel data changed size=" + data.size());
            mAdapter.submitList(data);
        });

        // ripristino la selezione salvata in onSaveInstanceState
        mAdapter.getSelectionTracker().onRestoreInstanceState(savedInstanceState);

        // la prima volta carico tutti i dati
//        mComicsViewModel.setFilter(null);
        mComicsViewModel.useLastFilter();

        return view;
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        // ripristino lo stato del layout (la posizione dello scroll)
        // se non trovo savedInstanceState uso lo stato salvato nel view model
        if (savedInstanceState != null) {
            Objects.requireNonNull(mRecyclerView.getLayoutManager())
                    .onRestoreInstanceState(savedInstanceState.getParcelable(BUNDLE_COMICS_RECYCLER_LAYOUT));
        } else if (mComicsViewModel != null) {
            Objects.requireNonNull(mRecyclerView.getLayoutManager())
                    .onRestoreInstanceState(mComicsViewModel.states.getParcelable(BUNDLE_COMICS_RECYCLER_LAYOUT));
        }
    }

    @Override
    public void onPause() {
        super.onPause();
        if (mRecyclerView != null) {
            // visto che Navigation ricrea il fragment ogni volta (!)
            // salvo lo stato della lista nel view model in modo da poterlo recuperare se necessario
            //  in onViewCreated
            mComicsViewModel.states.putParcelable(BUNDLE_COMICS_RECYCLER_LAYOUT,
                    Objects.requireNonNull(mRecyclerView.getLayoutManager()).onSaveInstanceState());
        }
    }

    @Override
    public void onAttach(@NonNull Context context) {
        super.onAttach(context);
        if (context instanceof OnNavigationFragmentListener) {
            mListener = (OnNavigationFragmentListener) context;
        } else {
            throw new RuntimeException(context.toString()
                    + " must implement OnNavigationFragmentListener");
        }
    }

    @Override
    public void onDetach() {
        super.onDetach();
        mListener = null;
    }

    @Override
    public void onSaveInstanceState(@NonNull Bundle outState) {
        super.onSaveInstanceState(outState);
        if (mAdapter != null) {
            // ripristino le selezioni
            mAdapter.getSelectionTracker().onSaveInstanceState(outState);
            // salvo lo stato del layout (la posizione dello scroll)
            outState.putParcelable(BUNDLE_COMICS_RECYCLER_LAYOUT,
                    Objects.requireNonNull(mRecyclerView.getLayoutManager())
                            .onSaveInstanceState());
        }
    }

    @Override
    public void onCreateOptionsMenu(@NonNull Menu menu, @NonNull MenuInflater inflater) {
        inflater.inflate(R.menu.menu_comics_fragment, menu);
        super.onCreateOptionsMenu(menu, inflater);

        final MenuItem searchItem = menu.findItem(R.id.searchComics);
        final SearchView searchView = (SearchView) searchItem.getActionView();

        // onQueryTextSubmit non viene scatenato su query vuota, quindi non posso caricare tutti i dati

        // TODO: al cambio di configurazione (es orientamento) la query viene persa
        //  è il viewModel che deve tenere memorizzata l'ultima query,
        //  qua al massimo devo apri la SearchView e inizializzarla con l'ultima query da viewModel se non vuota

        if (!TextUtils.isEmpty(mComicsViewModel.getLastFilter())) {
            // lo faccio prima di aver impostato i listener così non scateno più nulla
            searchItem.expandActionView();
            searchView.setQuery(mComicsViewModel.getLastFilter(), false);
            searchView.clearFocus();

            // TODO: non funziona sulla navigazione (es apro il dettaglio di un comics filtrato),
            //  perché viene chiusa la searchView e scatenato onQueryTextChange con testo vuoto
            //  che mi serve così perché quando volutamente la chiudo voglio che il filtro venga pulito
        }

        searchView.setOnQueryTextListener(new SearchView.OnQueryTextListener() {
            @Override
            public boolean onQueryTextSubmit(String query) {
                LogHelper.d("onQueryTextSubmit");
                return true;
            }

            @Override
            public boolean onQueryTextChange(String newText) {
                LogHelper.d("filterName change " + newText);
                mComicsViewModel.setFilter(newText); // TODO: ok ma aggiungere debounce
                return true;
            }
        });
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        if (item.getItemId() == R.id.createNewComics) {
            final NavDirections directions = ComicsFragmentDirections
                    .actionDestComicsFragmentToComicsEditFragment()
                    .setComicsId(NEW_COMICS_ID)
                    .setSubtitle(R.string.title_comics_create);

            Navigation.findNavController(requireView()).navigate(directions);

            return true;
        }
        return super.onOptionsItemSelected(item);
    }

    private void showUndo(Long[] ids, int count) {
        if (mUndoSnackBar != null && mUndoSnackBar.isShown()) {
            LogHelper.d("UNDO: dismiss snack");
            mUndoSnackBar.dismiss();
        }

        // mostro messaggio per undo
        // creo la snackbar a livello di activity così non ho grossi problemi quando cambio fragment
        mUndoSnackBar = Snackbar.make(requireActivity().findViewById(android.R.id.content),
                getResources().getQuantityString(R.plurals.comics_deleted, count, count), 7_000)
                // con il pulsante azione ripristino gli elementi rimossi
                .setAction(android.R.string.cancel, v -> mComicsViewModel.undoRemoved())
                .addCallback(new BaseTransientBottomBar.BaseCallback<Snackbar>() {
                    @Override
                    public void onDismissed(Snackbar transientBottomBar, int event) {
                        // procedo alla cancellazione effettiva solo dopo il timeout
                        if (event == BaseTransientBottomBar.BaseCallback.DISMISS_EVENT_TIMEOUT) {
                            mComicsViewModel.deleteRemoved();
                            // elimino anche le immagini
                            // mi fido del fatto che ids contenga esattamente i comics rimossi con l'istruzione sopra
                            ImageHelper.deleteImageFiles(requireContext(), ids);
                        }
                        LogHelper.d("UNDO: dismissed event=%s", event);
                    }
                });
        LogHelper.d("UNDO: show snack");
        mUndoSnackBar.show();
    }
}

File: app/src/main/java/it/amonshore/comikkua/ui/comics/ComicsFragment.kt
package it.amonshore.comikkua.ui.comics

import android.content.Context
import android.os.Bundle
import android.text.TextUtils
import android.view.*
import androidx.appcompat.view.ActionMode
import androidx.appcompat.widget.SearchView
import androidx.core.view.MenuHost
import androidx.core.view.MenuProvider
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import androidx.lifecycle.Lifecycle
import androidx.navigation.NavDirections
import androidx.navigation.Navigation.findNavController
import androidx.recyclerview.widget.LinearLayoutManager
import com.bumptech.glide.Glide
import it.amonshore.comikkua.LogHelper
import it.amonshore.comikkua.R
import it.amonshore.comikkua.data.comics.Comics
import it.amonshore.comikkua.data.comics.ComicsWithReleases
import it.amonshore.comikkua.databinding.FragmentComicsBinding
import it.amonshore.comikkua.parcelable
import it.amonshore.comikkua.ui.*

//private const val BUNDLE_COMICS_LAST_QUERY = "bundle.comics.last.query"
private val ACTION_MODE_NAME = ComicsFragment::class.java.simpleName + "_actionMode"
private const val BUNDLE_COMICS_RECYCLER_LAYOUT = "bundle.comics.recycler.layout"

class ComicsFragment : Fragment() {

    private val _viewModel: ComicsViewModel by viewModels()

    private lateinit var _listener: OnNavigationFragmentListener
    private lateinit var _adapter: PagedListComicsAdapter

    private var _binding: FragmentComicsBinding? = null
    private val binding get() = _binding!!

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        _binding = FragmentComicsBinding.inflate(layoutInflater, container, false)
        binding.list.layoutManager = LinearLayoutManager(requireContext())

        val actionModeController = createActionModeController()
        _adapter = createComicsAdapter(actionModeController)

        _viewModel.comicsWithReleasesPaged
            .observe(viewLifecycleOwner) { data ->
                LogHelper.d("comics viewmodel paging data changed")
                _adapter.submitData(lifecycle, data)
            }

        _viewModel.events
            .observe(viewLifecycleOwner) { result ->
                when (result) {
                    is UiComicsEvent.MarkedAsRemoved -> onMarkedAsRemoved(result.count, result.tag)
                }
            }

//        _comicsViewModelKt.useLastFilter()
        return binding.root
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        setupMenu()

        // ripristino lo stato del layout (la posizione dello scroll)
        // se non trovo savedInstanceState uso lo stato salvato nel view model
        if (savedInstanceState != null) {
            binding.list.layoutManager?.onRestoreInstanceState(
                savedInstanceState.parcelable(
                    BUNDLE_COMICS_RECYCLER_LAYOUT
                )
            )
        } else {
            binding.list.layoutManager?.onRestoreInstanceState(
                _viewModel.states.parcelable(
                    BUNDLE_COMICS_RECYCLER_LAYOUT
                )
            )
        }
    }

    override fun onPause() {
        super.onPause()
        // visto che Navigation ricrea il fragment ogni volta (!)
        // salvo lo stato della lista nel view model in modo da poterlo recuperare se necessario
        //  in onViewCreated
        _viewModel.states.putParcelable(
            BUNDLE_COMICS_RECYCLER_LAYOUT,
            binding.list.layoutManager?.onSaveInstanceState()
        )
    }

    override fun onAttach(context: Context) {
        super.onAttach(context)
        _listener = if (context is OnNavigationFragmentListener) {
            context
        } else {
            throw RuntimeException("$context must implement OnNavigationFragmentListener")
        }
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        _binding?.run {
            outState.putParcelable(
                BUNDLE_COMICS_RECYCLER_LAYOUT,
                binding.list.layoutManager?.onSaveInstanceState()
            )
        }
    }

    private fun createActionModeController() =
        object : ActionModeController(R.menu.menu_comics_selected) {
            override fun onActionItemClicked(mode: ActionMode, item: MenuItem): Boolean {
                if (item.itemId == R.id.deleteComics) {
                    val tracker = _adapter.selectionTracker
                    val ids = tracker.selection.toList()
                    _viewModel.markAsRemoved(ids)
                    tracker.clearSelection()
                    return true
                }
                return false
            }

            override fun onDestroyActionMode(mode: ActionMode) {
                // action mode distrutta (anche con BACK, che viene gestito internamente all'ActionMode e non può essere evitato)
                _adapter.selectionTracker.clearSelection()
                super.onDestroyActionMode(mode)
            }
        }

    /**
     * TODO: selection changed viene scatenato due volte all'inizio:
     *  questo perché il tracker permette la selezione di più item trascinando la selezione
     */
    private fun createComicsAdapter(
        actionModeController: ActionModeController
    ) = PagedListComicsAdapter.Builder(binding.list)
        .withOnItemSelectedListener { _, size ->
            if (size == 0) {
                _listener.onFragmentRequestActionMode(ACTION_MODE_NAME)
            } else {
                _listener.onFragmentRequestActionMode(
                    ACTION_MODE_NAME,
                    getString(R.string.title_selected, size),
                    actionModeController
                )
            }
        }
        .withComicsCallback(object : PagedListComicsAdapter.ComicsCallback {
            override fun onComicsClick(comics: ComicsWithReleases) {
                val directions: NavDirections = ComicsFragmentDirections
                    .actionDestComicsToComicsDetailFragment(comics.comics.id)
                findNavController(requireView()).navigate(directions)
            }

            override fun onComicsMenuSelected(comics: ComicsWithReleases) {
                BottomSheetDialogHelper.show(
                    requireActivity(), R.layout.bottomsheet_comics,
                    comics.comics.toSharable()
                ) { id: Int ->
                    when (id) {
                        R.id.createNewRelease -> {
                            openNewRelease(binding.root, comics)
                        }
                        R.id.share -> {
                            requireActivity().share(comics.comics)
                        }
                        R.id.deleteComics -> {
                            val ids = listOf(comics.comics.id)
                            _viewModel.markAsRemoved(ids)
                        }
                        R.id.search_starshop -> {
                            requireActivity().shareOnStarShop(comics.comics)
                        }
                        R.id.search_amazon -> {
                            requireActivity().shareOnAmazon(comics.comics)
                        }
                        R.id.search_popstore -> {
                            requireActivity().shareOnPopStore(comics.comics)
                        }
                        R.id.search_google -> {
                            requireActivity().shareOnGoogle(comics.comics)
                        }
                    }
                }
            }
        })
        .withGlide(Glide.with(this))
        .build()

    private fun openNewRelease(view: View, comics: ComicsWithReleases) {
        val directions: NavDirections = ComicsFragmentDirections
            .actionDestComicFragmentToReleaseEditFragment(comics.comics.id)
            .setSubtitle(R.string.title_release_create)
        findNavController(view).navigate(directions)
    }

    private fun setupMenu() {
        val menuHost: MenuHost = requireActivity()
        menuHost.addMenuProvider(object : MenuProvider {
            override fun onCreateMenu(menu: Menu, menuInflater: MenuInflater) {
                menuInflater.inflate(R.menu.menu_comics_fragment, menu)

                val searchItem = menu.findItem(R.id.searchComics)
                val searchView = searchItem.actionView as SearchView

                // onQueryTextSubmit non viene scatenato su query vuota, quindi non posso caricare tutti i dati

                // TODO: al cambio di configurazione (es orientamento) la query viene persa
                //  è il viewModel che deve tenere memorizzata l'ultima query,
                //  qua al massimo devo apri la SearchView e inizializzarla con l'ultima query da viewModel se non vuota
                if (!TextUtils.isEmpty(_viewModel.lastFilter)) {
                    // lo faccio prima di aver impostato i listener così non scateno più nulla
                    searchItem.expandActionView()
                    searchView.setQuery(_viewModel.lastFilter, false)
                    searchView.clearFocus()

                    // TODO: non funziona sulla navigazione (es apro il dettaglio di un comics filtrato),
                    //  perché viene chiusa la searchView e scatenato onQueryTextChange con testo vuoto
                    //  che mi serve così perché quando volutamente la chiudo voglio che il filtro venga pulito
                }

                searchView.setOnQueryTextListener(object : SearchView.OnQueryTextListener {
                    override fun onQueryTextSubmit(query: String): Boolean {
                        return true
                    }

                    override fun onQueryTextChange(newText: String): Boolean {
                        _viewModel.filter = newText // TODO: ok ma aggiungere debounce
                        return true
                    }
                })
            }

            override fun onMenuItemSelected(menuItem: MenuItem): Boolean {
                when (menuItem.itemId) {
                    R.id.createNewComics -> {
                        val directions: NavDirections = ComicsFragmentDirections
                            .actionDestComicsFragmentToComicsEditFragment()
                            .setComicsId(Comics.NEW_COMICS_ID)
                            .setSubtitle(R.string.title_comics_create)
                        findNavController(requireView()).navigate(directions)
                        return true
                    }
                    R.id.openComicsSelector -> {
                        val directions: NavDirections = ComicsFragmentDirections
                            .actionDestComicFragmentToComicsSelectorFragment()
                        findNavController(requireView()).navigate(directions)
                        return true
                    }
                    else -> return false
                }
            }
        }, viewLifecycleOwner, Lifecycle.State.RESUMED)
    }

    private fun onMarkedAsRemoved(count: Int, tag: String) {
        _listener.handleUndo(
            resources.getQuantityString(R.plurals.comics_deleted, count, count),
            tag
        )
    }
}